literalai/exporter.py

```diff
@@ -0,0 +1,443 @@
+from datetime import datetime, timezone
+import json
+from opentelemetry.sdk.trace import ReadableSpan
+from opentelemetry.sdk.trace.export import SpanExporter, SpanExportResult
+from typing import Dict, List, Optional, Sequence, cast
+import logging
+
+
+from literalai.event_processor import EventProcessor
+from literalai.helper import utc_now
+from literalai.observability.generation import GenerationType
+from literalai.observability.step import Step, StepDict
+
+
+class LoggingSpanExporter(SpanExporter):
+    def __init__(
+        self,
+        logger_name: str = "span_exporter",
+        event_processor: Optional[EventProcessor] = None,
+    ):
+        self.logger = logging.getLogger(logger_name)
+        self.logger.setLevel(logging.INFO)
+        self.event_processor = event_processor
+
+        if not self.logger.handlers:
+            handler = logging.StreamHandler()
+            formatter = logging.Formatter(
+                "%(asctime)s - %(name)s - %(levelname)s - %(message)s"
+            )
+            handler.setFormatter(formatter)
+            self.logger.addHandler(handler)
+
+    def export(self, spans: Sequence[ReadableSpan]) -> SpanExportResult:
+        """Export the spans by logging them."""
+        try:
+            for span in spans:
+                if (
+                    span.attributes
+                    and span.attributes.get("gen_ai.request.model", None) is not None
+                    and self.event_processor is not None
+                ):
+                    step = self._create_step_from_span(span)
+                    self.event_processor.add_event(cast(StepDict, step.to_dict()))
+
+            return SpanExportResult.SUCCESS
+        except Exception as e:
+            self.logger.error(f"Failed to export spans: {e}")
+            return SpanExportResult.FAILURE
+
+    def shutdown(self):
+        """Shuts down the exporter."""
+        if self.event_processor is not None:
+            return self.event_processor.flush_and_stop()
+
+    def force_flush(self, timeout_millis: float = 30000) -> bool:
+        """Force flush the exporter."""
+        return True
+
+    #     # TODO: Add generation promptid
+    #     # TODO: Add generation variables
+    #     # TODO: Check missing variables
+    #     # TODO: ttFirstToken
+    #     # TODO: duration
+    #     # TODO: tokenThroughputInSeconds
+    #     # TODO: Add tools
+    #     # TODO: error check with gemini error
```

Comment:
Haven't found a proper way to handle those as of now:
- promptid and variables could be set manually?
- missing variables like `frequency_penalty` are just not tracked by Traceloop so they do not appear in the span attributes, could only found basic support in their issue.
- ttFirstToken, duration and tokenThroughputInSeconds could not find a clear alternative
- tools to test

Reply:
Could not find a way to handle gemini and ttFirstToken